sysv-pdp11_man/u_man/man1/hpio.1
.TH HPIO 1 "3B20S only"
.SH NAME
hpio \- \s-1HP\s+1 2645A terminal tape file archiver
.SH SYNOPSIS
.B hpio
.BR \-o [ rc ]
file ...
.PP
.B hpio
.BR \-i [ rta ]
.RB [ \-n
count]
.SH DESCRIPTION
.I Hpio
is designed to take advantage of the tape drives on Hewlett Packard 2645A
terminals.
Up to 255 \s-1UNIX\s+1 System files can be archived onto a tape cartridge for off-line
storage or for transfer to another \s-1UNIX\s+1 System.
The actual number of files depends on the sizes of the files.
One file of about 115,000 bytes will almost fill a tape cartridge.
Almost 300 1-byte files will fit on a tape,
but the terminal will not be able to retrieve files after the first 255.
This manual page is not intended to be a guide for using tapes on
.SM HP
2645A terminals,
but tries to give enough information to be able to create and
read tape archives and to position a tape for access to a desired file in
an archive.
.PP
.B Hpio \-o
(copy out)
copies the specified
.IR file (s),
together with path name and status information to a tape drive on your
terminal
(which is assumed to be positioned at the beginning of a tape or immediately
after a tape mark).
The left tape drive is used by default.
Each
.I file
is written to a separate tape file and terminated with a tape mark.
When
.I hpio
finishes,
the tape is positioned following the last tape mark written.
.PP
.B Hpio \-i
(copy in)
extracts a file(s) from a tape drive
(which is assumed to be positioned at the beginning of a file that was
previously written by a
.BR "hpio \-o" ).
The default action extracts the next file from the left tape drive.
.PP
.I Hpio
always leaves the tape positioned after the last file read from or written
to the tape.
Tapes should always be rewound before the terminal is turned off.
To rewind a tape depress the green function button,
then function key 5,
and then select the appropriate tape drive by depressing either function key 5
for the left tape drive or function key 6 for the right.
If several files have been archived onto a tape,
the tape may be positioned at the beginning of a specific file by
depressing the green function button,
then function key 8,
followed by typing the desired file number (1\-255) with no \s-1RETURN\s+1,
and finally function key 5 for the left tape or function key 6 for the right.
The desired file number
may also be specified by a signed number relative to the current
file number.
.PP
The meanings of the available options are:
.PP
.PD 0
.TP \w'\fBn\fP\ \fIcount\fP\ \ 'u
.B r
Use the right tape drive.
.TP
.B c
Include a checksum at the end of each
.IR file .
The checksum is always checked by
.B hpio \-i
for each file written with this option by
.B hpio
.BR \-o .
.TP
.BI n " count"
The number of input files to be extracted is set to
.IR count .
If this option is not given,
.I count
defaults to 1.
An arbitrarily large
.I count
may be specified to extract all files from the tape.
.I Hpio
will stop at the end of data mark on the tape.
.TP
.B t
Print a table of contents only.
No files are created.
Printed information gives the file size in bytes,
the file name,
the file access modes,
and whether or not a checksum is included for the file.
.TP
.B a
Ask before creating a file.
.B Hpio \-i
normally prints the file size and name,
creates and reads in the file,
and prints a status message when the file has been read in.
If a checksum is included with the file,
it reports whether the checksum matched its computed value.
With this option,
the file size and name are printed followed by a
.BR ? .
Any response beginning with
.B y
or
.B Y
will cause the file to be copied in as above.
Any other response will cause the file to be skipped.
.PD
.SH FILES
.TP
/dev/tty??
to block messages while accessing a tape
.SH SEE ALSO
.IR "2645A Display Station User's Manual" ,
Hewlett-Packard Company,
Part Number 02645-90001.
.SH DIAGNOSTICS
.PD 0
.TP
.SM
BREAK
.br
An interrupt signal terminated processing.
.TP
Can't create `\fIfile\fP'.
File system access permissions did not allow
.I file
to be created.
.TP
Can't get tty options on stdout.
.I Hpio
was unable to get the input-output control settings associated with
the terminal.
.TP
Can't open `\fIfile\fP'.
.I File
could not be accessed to copy it to tape.
.TP
End of Tape.
No tape record was available when a read from a tape was requested.
An end of data mark is the usual reason for this,
but it may also occur if the wrong tape drive is being accessed and no
tape is present.
.TP
`\fIfile\fP' not a regular file.
.I File
is a directory or other special file.
Only regular files will be copied to tape.
.TP
Readcnt = \fIrc\fP, termcnt = \fItc\fP.
.I Hpio
expected to read
.I rc
bytes from the next block on the tape,
but the block contained
.I tc
bytes.
This is caused by having the tape improperly positioned or by a tape block
being mangled by interference from other terminal I/O.
.TP
Skip to next file failed.
An attempt to skip over a tape mark failed.
.TP
Tape mark write failed.
An attempt to write a tape mark at the end of a file failed.
.TP
Write failed.
A tape write failed.
This is most frequently caused by specifying the wrong tape drive,
running off the end of the tape,
or trying to write on a tape that is write protected.
.PD
.SH WARNINGS
Tape I/O operations may copy bad data if any other I/O involving the
terminal occurs.
Do not attempt any type ahead while
.I hpio
is running.
.I Hpio
turns off write permissions for other users while it is running,
but processes started asynchronously from your terminal can still interfere.
The most common indication of this problem,
while a tape is being written,
is the appearance of characters on the display screen that should have been
copied to tape.
.PP
The keyboard,
including the terminal's
.SM BREAK
key,
is locked during tape write operations;
the
.SM BREAK
key is only functional between writes.
.PP
.I Hpio
must have complete control of the attributes of the terminal to
communicate with the tape drives.
Interaction with commands such as
.IR cu (1C)
may interfere and prevent successful operation.
.SH BUGS
Some binary files contain sequences that will confuse the terminal.
.PP
An
.B hpio \-i
that encounters the end of data mark on the tape
(e.g., scanning the entire tape with \fBhpio \-itn 300\fP),
leaves the tape positioned \fIafter\fP the
end of data mark.
If a subsequent
.B hpio \-o
is done at this point,
the data will not be retrievable.
The tape must be repositioned manually using the terminal's
.SM FIND FILE\*S \-1
operation
(depress the green function button,
function key 8,
and then function key 5 for the left tape or
function key 6 for the right tape)
before the
.B hpio \-o
is started.
.PP
If an interrupt is received by
.I hpio
while a tape is being written,
the terminal may be left with the keyboard locked.
If this happens,
the terminal's
.SM RESET TERMINAL
key will unlock the keyboard.
.\"	@(#)hpio.1	5.2 of 5/18/82


sysvr3/301/usr/src/cmd/hpio/hpio.c
/*	Copyright (c) 1984 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF AT&T	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)hpio:hpio.c	1.2"
#include	"hpio.h"
#include	"stdio.h"
#include	"termio.h"
#include	"signal.h"
#include	"sys/types.h"
#include	"sys/stat.h"

#define	ENABLE	'\021'	/* Enable character. (DC1) */
#define	ENQ	'\005'	/* Enquiry character. (ENQ) */
#define	RS	'\036'	/* Record separator. (RS) */

/*
**	The "1" in Skip, Td, and Ts means left tape unit, changing it to 
**	"2" means right tape unit.
*/

char	Skip[] = "\033&p1u2C",	/* Tape file skip sequence. */
	Td[] = "\033&p1D",	/* Tape destination selection sequence. */
	Ts[] = "\033&p1S",	/* Tape source selection sequence. */
	Sreset[] = "\033g",	/* Soft reset of terminal. */
	check,			/* Checksum. */
	*drive = "left",	/* Selected tape drive name. */
	opts[] = "acin:ort";	/* Valid option letters. */

struct	FH	Fh;	/* File header. */
struct	stat	statbuf;/* Fstat buffer. */
struct	termio	Oterm,	/* Original terminal settings. */
		Nterm;	/* New terminal settings. */
int		aflg,	/* ask before input */
		cflg,	/* checksum */
		iflg,	/* input from tape */
		ncount,	/* number of input files */
		nflg,	/* use count of input files */
		oflg,	/* output to tape */
		tflg,	/* table of contents */	
		errflg,	/* option error detected */
		mode,	/* file access modes */
		ttymode;/* terminal access modes */
char		*ttynam;/* terminal file name */
extern int	optind;	/* argv index from getopt */
extern char	*optarg;/* option ptr from getopt */

int	catch();
char	*ttyname();

main(argc, argv)
int	argc;	/* arg count */
char	*argv[];/* arg vector */
{
	register FILE	*fp;	/* i/o file pointer */
	register char	c,	/* input character */
			*p;	/* ptr to buffer being checksummed */
	register int	i,	/* loop control */
			size2;	/* current record size */
	long		size;	/* current file size */
	char		buf[256];	/* i/o buffer */

	/* Accumulate options. */
	while((i = getopt(argc, argv, opts)) != EOF)
	switch(i) {
	case 'a':
		aflg = 1;
		break;
	case 'c':
		cflg = 'c';
		break;
	case 'i':
		iflg = 1;
		if(!nflg)
			ncount = 1;
		break;
	case 'n':
		nflg = 1;
		sscanf(optarg, "%d", &ncount);
		if(ncount <= 0)
			errflg++;
		break;
	case 'o':
		oflg = 1;
		break;
	case 'r':
		drive = "right";
		Td[3] = Ts[3] = Skip[3] = '2';
		break;
	case 't':
		tflg = 1;
		break;
	case '?':
		errflg++;
		break;
	}
	if(errflg || !(iflg ^ oflg)) {
		printf("Usage:\t%s\n\t%s\n",
			"hpio -i[art] [-n count]",
			"hpio -o[cr] file ...");
		exit(1);
	}

	/* Set up unbufferred raw mode with no echo. */
	setbuf(stdout, NULL);
	if(ioctl(fileno(stdout), TCGETA, &Oterm)) {
		printf("Can't get tty options on stdout.\n");
		exit(1);
	}
	ttynam = ttyname(fileno(stdout));
	fstat(fileno(stdout), &statbuf);
	ttymode = statbuf.st_mode;
	chmod(ttynam, ttymode & ~077);
	signal(SIGINT, catch);
	Nterm.c_iflag = BRKINT;
	Nterm.c_oflag = 0;
	Nterm.c_cflag = (Oterm.c_cflag & ~PARENB) | CS8;
	Nterm.c_lflag = NOFLSH;
	Nterm.c_line = 0;
	Nterm.c_cc[VMIN] = 6;
	Nterm.c_cc[VTIME] = 1;
	ioctl(fileno(stdout), TCSETAW, &Nterm);

	/* Determine desired action and perform it. */
	if(iflg) {

		/* Select source tape and loop through selected files. */
		for(printf("%s", Ts);ncount--;hskip()) {

			/* Read the header structure from the tape. */
			hread(&Fh, sizeof(Fh));
			sscanf(Fh.FHsize, SIZEFMT, &size);
			printf("%ld bytes of `%s'", size, Fh.FHname);
			sscanf(Fh.FHmode, MODEFMT, &mode);
			if(tflg) {

				/* Table of contents only. */
				printf(" mode: %#o with%s checksum.\r\n", mode,
					Fh.FHchkf ? "" : "out");
				continue;
			}
			if(aflg) {

				/* Ask before copying file in. */
				printf("? ");
				c = getchar();
				if(c != 'Y' && c != 'y') {
					putchar(c);
					while(c != '\r')
						putchar(c = getchar());
					putchar('\n');
					continue;
				}
				while(c != '\r') {
					putchar(c);
					c = getchar();
				}
			}
			printf("\r\n");

			/* Create the file to be read in. */
			if((fp = fopen(Fh.FHname, "w")) == NULL) {
				printf("Can't create `%s'.\r\n", Fh.FHname);
				continue;
			}

			/* Copy the file from the tape to disk. */
			check = 0;
			while(size > 0) {
				size2 = size > 256 ? 256 : size;
				hread(buf, size2);
				fwrite(buf, 1, size2, fp);
				if(Fh.FHchkf)
					for(i = size2, p = buf;i--;
						check ^= *p++);
				size -= size2;
			}
			fclose(fp);
			printf("`%s' read from %s tape.\r\n",
				Fh.FHname, drive);
			if(Fh.FHchkf) {

				/* Check checksum. */
				hread(buf, 5);
				sscanf(buf, CHKFMT, &errflg);
				printf("Checksum match %s.\r\n",
					(errflg == (check & 0377)) ?
					"succeeded" : "\007failed");
			}
			if(mode)
				chmod(Fh.FHname, mode);
		}
	} else {

		/* Select destination tape and copy files to tape. */
		for(printf("%s", Td), i = optind;i < argc;i++) {

			/* Setup file header and open the input file. */
			strcpy(Fh.FHname, argv[i]);
			Fh.FHchkf = cflg;
			check = 0;
			if((fp = fopen(Fh.FHname, "r")) == NULL) {
				printf("Can't open `%s'.\r\n", Fh.FHname);
				continue;
			}
			fstat(fileno(fp), &statbuf);
			if((statbuf.st_mode & S_IFMT) != S_IFREG) {
				printf("`%s' not a regular file.\r\n",
					Fh.FHname);
				fclose(fp);
				continue;
			}
			sprintf(Fh.FHmode, MODEFMT, statbuf.st_mode & 07777);
			sprintf(Fh.FHsize, SIZEFMT, size = statbuf.st_size);

			/* Write header record. */
			hwrite(&Fh, sizeof(Fh));

			/* Write file. */
			while(size > 0) {
				size2 = size > 256 ? 256 : size;
				if(fread(buf, 1, size2, fp) != size2) {
					sprintf(buf, "Read fail on `%s'.",
						Fh.FHname);
					exitp(buf);
				}
				hwrite(buf, size2);
				size -= size2;
				if(cflg)
					for(p = buf;size2--;check ^= *p++);
			}
			fclose(fp);

			/* Write checksum. */
			if(cflg) {
				sprintf(buf, CHKFMT, check & 0377);
				hwrite(buf, 5);
			}
			hmark();
			printf("%s bytes of `%s' written to %s tape.\r\n",
				Fh.FHsize, Fh.FHname, drive);
		}
	}
	ioctl(fileno(stdout), TCSETAW, &Oterm);
	chmod(ttynam, ttymode);
	exit(0);
}

/*
**	hwrite - Write to HP cassette drive.
**
**	Write a record of the given size from the given output buffer.
*/

hwrite(buf, cnt)
char	*buf;	/* ptr to output buffer */
int	cnt;	/* # of bytes to write */
{

	/* Put out write command and get go ahead signal. */
	printf("\033&p%dW%c", cnt, ENQ);
	getchar();

	/* Write the data and get completion status. */
	fwrite(buf, 1, cnt, stdout);
	putchar(ENABLE);
	if(getchar() != 'S')
		exitp("Write failed.");
	getchar();
	return(cnt);
}

/*
**	hskip - Skip to start of next file on HP cassette drive.
**
**	This routine requests a skip to the next file on the cassette and
**	waits for completion of the operation.
*/

hskip()
{
	printf("%s%c", Skip, ENABLE);
	if(getchar() != 'S')
		exitp("Skip to next file failed.");
	getchar();
}

/*
**	hread - Read from HP cassette drive.
**
**	Read the next record from the current file on the tape drive.
**	It must be the the same size as requested.
*/

hread(buf, cnt)
register char	*buf;	/* ptr to buffer to receive the record */
int		cnt;	/* size of desired record */
{
	register int	i,	/* loop control */
			cnt2;	/* # of bytes left to be read */
	register char	c;	/* input character */

	/* Request record read and get byte count for the record. */
	printf("\033&p2R%c", ENABLE);
	for(cnt2 = 0, i = 4;i--;cnt2 = cnt2 * 16 + c - '0')
		if((c = getchar()) == RS)
			exitp("End of Tape.");
	getchar();
	if(cnt2 != cnt) {
		sprintf(buf, "Readcnt = %d, termcnt = %d.", cnt, cnt2);
		exitp(buf);
	}

	/* Enable data record transmission and read it into buf. */
	putchar(ENABLE);
	for(;cnt2--;*buf++ = getchar());
	return(cnt);
}

/*
**	hmark - Write tape mark on HP cassette drive.
**
**	Write a tape mark and get the completion status of the operation.
*/

hmark()
{
	printf("\033&p5C%c", ENABLE);
	if(getchar() != 'S')
		exitp("Tape mark write failed.");
	getchar();
}

/*
**	catch	Catch break.
**
**	This routine catches the interrupt signal to reset terminal
**	options and exit.
*/

catch()
{
	signal(SIGINT, SIG_IGN);
	exitp("BREAK");
}

/*
**	exitp - Error exit.
**
**	Reset terminal modes, print error message, and exit.
*/

exitp(mp)
char	*mp;	/* ptr to error message */
{
	printf("%s", Sreset);
	sleep(2);
	printf("%s\r\n", mp);
	ioctl(fileno(stdout), TCSETAW, &Oterm);
	chmod(ttynam, ttymode);
	exit(1);
}

